caddyhttp/httpserver/server.go

```diff
@@ -265,6 +269,14 @@ func (s *Server) serveHTTP(w http.ResponseWriter, r *http.Request) (int, error)
 		}
 	}
 
+	// URL fields other than Path and RawQuery will be empty for most server
+	// requests. Hence, the request URL is updated with the scheme and host
+	// from the virtual host's site address.
+	if vhostURL, err := url.Parse(vhost.Addr.String()); err == nil {
+		r.URL.Scheme = vhostURL.Scheme
+		r.URL.Host = vhostURL.Host
```

Comment:
The URL in request r has neither the scheme nor the host.

http.Redirect will [set the Location header to a path](https://github.com/golang/go/blob/64c02460d703f718a44326646d19c417eeb1999a/src/net/http/server.go#L2011) (not absolute URI) when
the provided urlStr is missing both.

Since the virtual host's site address has both, and the virtual host
configuration was matched to r.URL.Path, r.URL is updated so that invoking
http.Redirect will:

1. Set the Location header to an absolute URI
2. Redirect to a URI with the correct scheme